python_modules/dagster/dagster/_core/definitions/time_window_partition_mapping.py

```diff
@@ -97,17 +99,72 @@ def get_upstream_partitions_for_partitions(
         self,
         downstream_partitions_subset: Optional[PartitionsSubset],
         upstream_partitions_def: PartitionsDefinition,
+        current_time: Optional[datetime] = None,
         dynamic_partitions_store: Optional[DynamicPartitionsStore] = None,
     ) -> PartitionsSubset:
+        """Returns the partitions in the upstream asset that map to the given downstream partitions.
+
+        Raises an error if upstream partitions do not exist at the given current_time, fetching the
+        current time if not provided.
+        """
         if not isinstance(downstream_partitions_subset, TimeWindowPartitionsSubset):
             check.failed("downstream_partitions_subset must be a TimeWindowPartitionsSubset")
 
-        return self._map_partitions(
-            downstream_partitions_subset.partitions_def,
+        if not isinstance(upstream_partitions_def, TimeWindowPartitionsDefinition):
```

Comment:
it's not clear to me why the get_upstream_partitions_for_partitions and get_downstream_partitions_for_partitions implementations need to diverge from _map_partitions here. It seems like they're still doing essentially the same thing as each other

Reply:
It's a small difference -- `get_upstream_partitions_for_partitions` raises errors if mapped time windows don't exist at the current time, while `get_downstream_partitions_for_partitions` just filters the time windows to only contain existent time windows.

I forked the implementations here because it felt better than passing a flag into `_map_partitions`, but I'm open to reworking this if you have recommendations

Reply:
I think passing a flag for "raise_on_non_existent" would be better/more maintainable here. As it stands, if we ever want to update the logic for _map_partitions to be (e.g.) more efficient, we'd have to update both of these places, which runs a pretty high risk of divergence